components/fxa-client/ios/FxAClient/FxAccountManager.swift

```diff
@@ -0,0 +1,584 @@
+
+                postAuthenticated(authType: .recovered)
+
+                return Event.fetchProfile
+            }
+            case .fetchProfile: do {
+                // Profile fetching and account authentication issues:
+                // https://github.com/mozilla/application-services/issues/483
+                FxALog.info("Fetching profile...")
+
+                do {
+                    profile = try handleAccountAuthErrors { acct in
+                        try acct.getProfileSync()
+                    }
+                } catch {
+                    return Event.failedToFetchProfile
+                }
+                return Event.fetchedProfile
+            }
+            default: break // Do nothing
+            }
+        }
+        case .authenticatedWithProfile: do {
+            switch via {
+            case .fetchedProfile: do {
+                DispatchQueue.main.async {
+                    NotificationCenter.default.post(
+                        name: .accountProfileUpdate,
+                        object: nil,
+                        userInfo: ["profile": self.profile!]
+                    )
+                }
+            }
+            default: break // Do nothing
+            }
+        }
+        case .authenticationProblem:
+            switch via {
+            case .authenticationError: do {
+                // Somewhere in the system, we've just hit an authentication problem.
+                // There are two main causes:
+                // 1) an access token we've obtain from fxalib via 'getAccessToken' expired
+                // 2) password was changed, or device was revoked
+                // We can recover from (1) and test if we're in (2) by asking the fxalib.
+                // If it succeeds, then we can go back to whatever
+                // state we were in before. Future operations that involve access tokens should
+                // succeed.
+
+                func onError() {
+                    // We are either certainly in the scenario (2), or were unable to determine
+                    // our connectivity state. Let's assume we need to re-authenticate.
+                    // This uncertainty about real state means that, hopefully rarely,
+                    // we will disconnect users that hit transient network errors during
+                    // an authorization check.
+                    // See https://github.com/mozilla-mobile/android-components/issues/3347
+                    FxALog.error("Unable to recover from an auth problem.")
+                    DispatchQueue.main.async {
+                        NotificationCenter.default.post(
+                            name: .accountAuthProblems,
+                            object: nil
+                        )
+                    }
+                }
+
+                do {
+                    return try handleAccountAuthErrors { acct in
+                        let info = try acct.checkAuthorizationStatusSync()
```

Comment:
@rfk is this the correct way to use the auth status endpoint and to do a recovery flow in your mind?

Reply:
(e.g. does it throw the expected errors, or return some failure-success response)

Reply:
This looks correct to me 👍